### ISSUE TITLE
Cmake tools 1.4.0 extension is stuck at scanning for kits. 

### ISSUE THREAD

**Issue by sourabh2902** [microsoft/vscode-cmake-tools#1289]
### Brief Issue Summary

Hi, 
I am a new user of VS Code and CMAKE TOOLS, hence kindly give me some help for the silly issue that I am facing from some time. I am working on large C++ based project, which is uses CMAKE. When I install the latest CMAKE TOOLS extension, it is just stuck at "Scanning for Kits" stage and doesn't allow me to configure my project further. I don't see the same issue with older versions of this extension. 

### Expected:
On the older version (1.3.0) the steps I did were: 
1. Install Cmake tools, 
2. open workspace, 
3. choose [unspecified kit] from command pallete
4. select variant and configure. 
5. Build the target and run tests. 

### Apparent Behavior:
In the version 1.4.0 I see the following : 
1. Install Cmake tools, 
2. open workspace, 
3. Tool popup shows "scanning for kits .......... " and this never finishes . 
4. Tool doesnt allow me to select a variant and proceed further. 

### CMake Tools Log
Please note I have replaced long path names with <some_path> in the log below 

2020-06-07T08:34:50.918Z [debug] [kit] Detected GCC compiler: /usr/bin/X11/gcc-4.8
2020-06-07T08:34:50.918Z [debug] [kit] Detected GCC compiler: /usr/bin/X11/gcc
2020-06-07T08:34:50.919Z [debug] [kit] Detected GCC compiler: /usr/bin/gcc-4.8
2020-06-07T08:34:50.920Z [debug] [kit] Detected GCC compiler: /usr/bin/gcc
2020-06-07T08:34:50.921Z [debug] [kit] Detected Clang compiler: <some_path>/YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin/clang-9
2020-06-07T08:34:50.921Z [debug] [kit] Detected Clang compiler: <some_path>/YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin/clang
2020-06-07T08:34:50.922Z [debug] [kit] Detected GCC compiler: <some_path>/gnu64/gcc/7.2.0/bin/x86_64-pc-linux-gnu-gcc-7.2.0
2020-06-07T08:34:50.922Z [debug] [kit] Detected GCC compiler: <some_path>/gnu64/gcc/7.2.0/bin/x86_64-pc-linux-gnu-gcc
2020-06-07T08:34:50.923Z [debug] [kit] Detected GCC compiler: <some_path>/gnu64/gcc/7.2.0/bin/gcc
2020-06-07T08:34:50.950Z [debug] [kit] Found 2 kits in directory /usr/bin/X11
2020-06-07T08:34:50.950Z [debug] [kit] Found 2 kits in directory /usr/bin
2020-06-07T08:34:50.951Z [debug] [kit] Found 2 kits in directory /prj/vlsi/mdsp/fw_public/YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin
2020-06-07T08:34:50.951Z [debug] [kit] Found 3 kits in directory /pkg/qct/software/gnu64/gcc/7.2.0/bin
2020-06-07T08:34:50.970Z [debug] [kit] Bad GCC binary /pkg/gnu/bin/i686-pc-linux-gnu-gcc -v output: Reading specs from /pkg/gnu/bin/../lib/gcc-lib/i686-pc-linux-gnu/3.2/specs
Configured with: ../../gcc-3.2/configure --prefix=/pkg/gnu/gcc/3.2
Thread model: posix
gcc version 3.2

2020-06-07T08:34:50.970Z [debug] [kit] Bad GCC binary /pkg/gnu/bin/gcc -v output: Reading specs from /pkg/gnu/bin/../lib/gcc-lib/i686-pc-linux-gnu/3.2/specs
Configured with: ../../gcc-3.2/configure --prefix=/pkg/gnu/gcc/3.2
Thread model: posix
gcc version 3.2

2020-06-07T08:34:50.970Z [debug] [kit] Found 0 kits in directory /pkg/gnu/bin

### Developer Tools Log
Sorry I am not able to find this log 
### Platform and Versions

- **Operating System**: 
- **CMake Version**: 
- **VSCode Version**: 
- **CMake Tools Extension Version**: 
- **Compiler/Toolchain**: 

### Other Notes/Information
for the development I am interested to use, the following 
2020-06-07T08:34:50.951Z [debug] [kit] Found 2 kits in directory /<some_path>/YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin
2020-06-07T08:34:50.951Z [debug] [kit] Found 3 kits in directory <some_path>/gnu64/gcc/7.2.0/bin

**Comment by andreeis** [comment]
You can get unblocked by creating a new simple CMake project (with the cmake.QuickStart command) which will not have this problem and let that one scan for kits. It should not manifest the problem. The second time you open any project with CMake Tools 1.4.0, this silent scanning for kits will not happen (because it is triggered by detecting a version change in the kits definition) and you should be able to continue as before.

But it would be very useful to get as much information as possible about this problem before you get passed the problem state. It will not be straightforward that, once you get unblocked, to get back to the buggy state (moving to 1.3.1 and back to 1.4.0 won't do it).

Besides the kits json file in the user path, do you also have an extra kits json in the .vscode subfolder? Can you share the content of the kits json file(s) as it is now, in the buggy state?

We have a bug in 1.4.0 regarding exactly scanning for kits, but it should not manifest with a hang like that, that is why I am asking for more information. We will release an update 1.4.1 soon.

**Comment by sourabh2902** [comment]
Hi @andreeis  
Thanks for giving my issue priority, Kindly take a look at my observation below : 

**Workaround the problem using  Quick Start :**    
Let's assume that I have presently installed 1.3.0 to use this workaround I am planning to follow these steps, please correct me if you see any mistakes: 

1. Uninstall 1.3.0 
2. Clear the build area 
3. Close vscode 
4.Reopen vscode , 
5. Install 1.4.0 
6. Don't open workspace ! as that will trigger the silent scanning , instead use cmake:quick start 

**Kits.json file**
In version 1.3.0
1. I didn't have any kits.json file, neither in .vscode/ nor in /usr/.local/share/CMakeTools/ 
2. I added a kits json file to .vscode/ just to check if it gets picked up by the tool  (1.3.0) . 
3. I saw the following message 👍 
2020-06-08T18:45:23.889Z [info] [kit] Successfully loaded 1 kits from <some_path>/ide/.vscode/cmake-kits.json
2020-06-08T18:45:23.890Z [debug] [main] Injecting new Kit into CMake driver
4. The content of cmake-kits.json was : 
[
  {
    "name":"GCC7.2.0",
    "compilers": {
      "CC":"/pkg/qct/software/gnu64/gcc/7.2.0/bin/gcc",
      "CXX":"/pkg/qct/software/gnu64/gcc/7.2.0/bin/g++"
    }
  }
]

In version 1.4.0
1. I used the same cmake-kits.json from the step above at the same location, with the same content
2. The tool did not pick up the file, kept on scanning for kits. 
3. For this exercise ,to go from 1.3.0 to 1.4.0 I followed these steps -

- Uninstall 1.3.0 
-  Clear the build area 
-  Close vscode 
- Reopen vscode , 
-  Install 1.4.0 
- Clicked on open workspace and continued ....

**Comment by andreeis** [comment]
Whether to scan for kits or not is tied to a global state variable that is not impacted by deleting the build folder, nor by moving back and forth between releases. Not even by the command "reset state (for troubleshooting)", which resets any local workspace state variables. I am not yet aware of a way to reset the global state variable, unless with cloning CMake Tools,  changing the code and running the "extension development host".

Anyway, good clue that you didn't have any kits json file at all, I will play with that.

In 1.4.0, running the quickStart will trigger the silent scanning only once, the first time. In 1.3.0, quickStart will not trigger silent scanning.

When you say "The tool did not pick up the file, kept on scanning for kits." you mean that it hanged again or just the fact that the scan happened? The scan is supposed to happen even if you have both files. It may make updates in its content. So let me know if that phrase means it just scanned or it still hanged.

**Comment by sourabh2902** [comment]
Hi @andreeis  
I meant the tool still hanged . 
It just showed a pop up  Scanning for kits .... which never finished , in the log files there were no changes either(exactly as I showed in my first message)

**Comment by andreeis** [comment]
Just to have as many clues as possible, share also the contents of .vscode\settings.json and your user level settings.json. 
And you said that originally you had no user kit file in /usr/.local/share/CMakeTools/. but the hang still occurs even if you have that file created (by the command 'scan for kits')? You must have that file now present in that location, right?

**Comment by sourabh2902** [comment]
HI @andreeis 
Please take a look below : 

**.vscode/settings.json**
cmake.cmakePath":"cmake",
    "cmake.sourceDirectory": "${env:WORKSPACE}/model/arch",
    "cmake.buildDirectory": "${env:UNMANAGED_DIR}/cmake",

    "cmake.generator": "Ninja",



    "C_Cpp.autocomplete": "Disabled",
    "C_Cpp.formatting": "Disabled",
    "C_Cpp.errorSquiggles": "Disabled",
    "C_Cpp.intelliSenseEngine": "Disabled",
    "python.pythonPath": "<some_path>anaconda3/anaconda3/envs/anaconda3-py3.5-2018-01/bin/python3"

**user level settings.json**
{
    "window.zoomLevel": 1,
    "workbench.colorTheme": "Monokai Dimmed",
    "editor.minimap.enabled": false,
    "workbench.sideBar.location": "right",
    "editor.tabSize": 2,
    "files.trimFinalNewlines": true,
    "files.trimTrailingWhitespace": true,
    "workbench.editor.enablePreview": false,
    "workbench.editor.enablePreviewFromQuickOpen": false,
    "workbench.iconTheme": "vscode-icons",
    "window.titleBarStyle": "custom",
    "C_Cpp.formatting": "Disabled",
    "C_Cpp.loggingLevel": "Debug",
    "python.dataScience.sendSelectionToInteractiveWindow": true,
    "editor.wordWrap": "wordWrapColumn",
    "editor.wordWrapColumn": 170
}

**.vscode\vscode_workspace.code_workspace*
{
	"folders": [
		<some dir paths ,> 
	],

	"settings": {
		"clangd.arguments": [
			"--compile-commands-dir=<some_path>/cmake"
		],

		"files.associations": {
			"optional": "cpp",
			"bitset": "cpp"
		},

		"C_Cpp.intelliSenseEngine": "Disabled",
		"C_Cpp.intelliSenseEngineFallback": "Disabled",
		"cmake.autoSelectActiveFolder": false,
		"cmake.configureOnOpen": false,
		"cmake.loggingLevel": "debug"
	},
}

**log for ~/.local/share/CmakeTools/log.txt**
2020-06-15T02:12:58.928Z [debug] [kit] Scanning for Kits on system
2020-06-15T02:12:58.934Z [debug] [kit] Scanning directory /pkg/software/kcachegrind/bin for compilers
2020-06-15T02:12:58.934Z [debug] [kit] Scanning directory /pkg/software/gnu64/gcc/7.2.0/bin for compilers
2020-06-15T02:12:58.934Z [debug] [kit] Scanning directory /prj//YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin for compilers
2020-06-15T02:12:58.935Z [debug] [kit] Scanning directory /prj/local/anaconda3/bin for compilers
2020-06-15T02:12:58.935Z [debug] [kit] Scanning directory /pkg/software/ninja/1.9.0 for compilers
2020-06-15T02:12:58.935Z [debug] [kit] Scanning directory /pkg/software/cmake/3.12.2/bin for compilers
2020-06-15T02:12:58.936Z [debug] [kit] Scanning directory /pkg/software/gnu64/binutils/2.25/bin for compilers
2020-06-15T02:12:58.936Z [debug] [kit] Scanning directory /pkg/software/gnu/gdb/7.9.1/bin for compilers
2020-06-15T02:12:58.937Z [debug] [kit] Scanning directory /pkg/software/gnu/ddd/3.3.12/bin for compilers
2020-06-15T02:12:58.937Z [debug] [kit] Scanning directory /pkg/software/valgrind/3.9.0/bin for compilers
2020-06-15T02:12:58.937Z [debug] [kit] Scanning directory /pkg/software/python/2.7.9/bin for compilers
2020-06-15T02:12:58.937Z [debug] [kit] Scanning directory /pkg/qvmr/bin for compilers
2020-06-15T02:12:58.941Z [debug] [kit] Scanning directory /pkg/bin for compilers
2020-06-15T02:12:58.941Z [debug] [kit] Scanning directory /pkg/wts/bin for compilers
2020-06-15T02:12:58.941Z [debug] [kit] Scanning directory /usr/bin for compilers
2020-06-15T02:12:58.942Z [debug] [kit] Scanning directory /bin for compilers
2020-06-15T02:12:58.942Z [debug] [kit] Scanning directory /usr/sbin for compilers
2020-06-15T02:12:58.942Z [debug] [kit] Scanning directory /sbin for compilers
2020-06-15T02:12:58.943Z [debug] [kit] Scanning directory /usr/local/bin for compilers
2020-06-15T02:12:58.943Z [debug] [kit] Scanning directory /usr/bin/X11 for compilers
2020-06-15T02:12:58.943Z [debug] [kit] Skipping scan of not existing path /usr/X11R6/bin
2020-06-15T02:12:58.943Z [debug] [kit] Found 0 kits in directory /usr/X11R6/bin
2020-06-15T02:12:58.944Z [debug] [kit] Scanning directory /usr/games for compilers
2020-06-15T02:12:58.944Z [debug] [kit] Skipping scan of not existing path /opt/kde3/bin
2020-06-15T02:12:58.944Z [debug] [kit] Found 0 kits in directory /opt/kde3/bin
2020-06-15T02:12:58.945Z [debug] [kit] Scanning directory /usr/lib/mit/bin for compilers
2020-06-15T02:12:58.945Z [debug] [kit] Scanning directory /usr/lib/mit/sbin for compilers
2020-06-15T02:12:58.945Z [debug] [kit] Scanning directory /pkg/icetools/bin for compilers
2020-06-15T02:12:58.946Z [debug] [kit] Scanning directory /pkg/netscape/bin for compilers
2020-06-15T02:12:58.946Z [debug] [kit] Scanning directory /pkg/gnu/bin for compilers
2020-06-15T02:12:58.946Z [debug] [kit] Scanning directory /pkg/qcaetools/syseng/bin for compilers
2020-06-15T02:12:58.948Z [debug] [kit] Skipping scan of not existing path /local/mnt/workspace/eclipse/eclipse
2020-06-15T02:12:58.948Z [debug] [kit] Found 0 kits in directory /local/mnt/workspace/eclipse/eclipse
2020-06-15T02:12:58.949Z [debug] [kit] Skipping scan of not existing path /local/mnt/workspace/jre/jre1.5.0_17/bin
2020-06-15T02:12:58.949Z [debug] [kit] Found 0 kits in directory /local/mnt/workspace/jre/jre1.5.0_17/bin
2020-06-15T02:12:58.949Z [debug] [kit] Scanning directory /pkg/software/tkdiff/current for compilers
2020-06-15T02:12:58.957Z [debug] [kit] Found 0 kits in directory /pkg/software/kcachegrind/bin
2020-06-15T02:12:58.958Z [debug] [kit] Testing GCC binary: /pkg/software/gnu64/gcc/7.2.0/bin/gcc
2020-06-15T02:12:58.972Z [debug] [kit] Testing GCC binary: /pkg/software/gnu64/gcc/7.2.0/bin/x86_64-pc-linux-gnu-gcc
2020-06-15T02:12:58.982Z [debug] [kit] Testing GCC binary: /pkg/software/gnu64/gcc/7.2.0/bin/x86_64-pc-linux-gnu-gcc-7.2.0
2020-06-15T02:12:58.993Z [debug] [kit] Testing Clang binary: /prj//YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin/clang
2020-06-15T02:12:58.994Z [debug] [kit] Testing Clang binary: /prj//YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin/clang
2020-06-15T02:12:59.007Z [debug] [kit] Testing Clang binary: /prj//YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin/clang-9
2020-06-15T02:12:59.007Z [debug] [kit] Testing Clang binary: /prj//YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin/clang-9
2020-06-15T02:12:59.043Z [debug] [kit] Found 0 kits in directory /prj/local/anaconda3/bin
2020-06-15T02:12:59.044Z [debug] [kit] Found 0 kits in directory /pkg/software/ninja/1.9.0
2020-06-15T02:12:59.045Z [debug] [kit] Found 0 kits in directory /pkg/software/cmake/3.12.2/bin
2020-06-15T02:12:59.046Z [debug] [kit] Found 0 kits in directory /pkg/software/gnu64/binutils/2.25/bin
2020-06-15T02:12:59.046Z [debug] [kit] Found 0 kits in directory /pkg/software/gnu/gdb/7.9.1/bin
2020-06-15T02:12:59.047Z [debug] [kit] Found 0 kits in directory /pkg/software/gnu/ddd/3.3.12/bin
2020-06-15T02:12:59.048Z [debug] [kit] Found 0 kits in directory /pkg/software/valgrind/3.9.0/bin
2020-06-15T02:12:59.057Z [debug] [kit] Found 0 kits in directory /pkg/software/python/2.7.9/bin
2020-06-15T02:12:59.113Z [debug] [kit] Found 0 kits in directory /pkg/qvmr/bin
2020-06-15T02:12:59.119Z [debug] [kit] Found 0 kits in directory /pkg/wts/bin
2020-06-15T02:12:59.145Z [debug] [kit] Found 0 kits in directory /pkg/bin
2020-06-15T02:12:59.153Z [debug] [kit] Found 0 kits in directory /bin
2020-06-15T02:12:59.166Z [debug] [kit] Found 0 kits in directory /sbin
2020-06-15T02:12:59.167Z [debug] [kit] Found 0 kits in directory /usr/local/bin
2020-06-15T02:12:59.196Z [debug] [kit] Found 0 kits in directory /usr/sbin
2020-06-15T02:12:59.197Z [debug] [kit] Found 0 kits in directory /usr/games
2020-06-15T02:12:59.198Z [debug] [kit] Found 0 kits in directory /usr/lib/mit/bin
2020-06-15T02:12:59.198Z [debug] [kit] Found 0 kits in directory /usr/lib/mit/sbin
2020-06-15T02:12:59.204Z [debug] [kit] Testing GCC binary: /pkg/icetools/bin/Powerpc-eabi-gcc
2020-06-15T02:12:59.268Z [debug] [kit] Testing GCC binary: /usr/bin/X11/gcc
2020-06-15T02:12:59.279Z [debug] [kit] Testing GCC binary: /usr/bin/X11/gcc-4.8
2020-06-15T02:12:59.393Z [debug] [kit] Found 0 kits in directory /pkg/netscape/bin
2020-06-15T02:12:59.394Z [debug] [kit] Testing GCC binary: /pkg/gnu/bin/gcc
2020-06-15T02:12:59.406Z [debug] [kit] Testing GCC binary: /pkg/gnu/bin/i686-pc-linux-gnu-gcc
2020-06-15T02:12:59.425Z [debug] [kit] Found 0 kits in directory /pkg/qcaetools/syseng/bin
2020-06-15T02:12:59.440Z [debug] [kit] Found 0 kits in directory /pkg/software/tkdiff/current
2020-06-15T02:12:59.570Z [debug] [kit] Testing GCC binary: /usr/bin/gcc
2020-06-15T02:12:59.582Z [debug] [kit] Testing GCC binary: /usr/bin/gcc-4.8
2020-06-15T02:12:59.699Z [debug] [kit] Detected Clang compiler: /prj//YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin/clang-9
2020-06-15T02:12:59.700Z [debug] [kit] Detected Clang compiler: /prj//YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin/clang2020-06-15T02:12:59.700Z [debug] [kit] Detected GCC compiler: /pkg/software/gnu64/gcc/7.2.0/bin/x86_64-pc-linux-gnu-gcc-7.2.0
2020-06-15T02:12:59.701Z [debug] [kit] Detected GCC compiler: /pkg/software/gnu64/gcc/7.2.0/bin/x86_64-pc-linux-gnu-gcc
2020-06-15T02:12:59.701Z [debug] [kit] Detected GCC compiler: /pkg/software/gnu64/gcc/7.2.0/bin/gcc
2020-06-15T02:12:59.707Z [debug] [kit] Found 2 kits in directory /prj//YCM/clang+llvm-9.0.0-x86_64-linux-sles11.3/bin
2020-06-15T02:12:59.707Z [debug] [kit] Found 3 kits in directory /pkg/software/gnu64/gcc/7.2.0/bin
2020-06-15T02:12:59.708Z [debug] [kit] Detected GCC compiler: /usr/bin/gcc-4.8
2020-06-15T02:12:59.709Z [debug] [kit] Detected GCC compiler: /usr/bin/gcc
2020-06-15T02:12:59.709Z [debug] [kit] Bad GCC binary /pkg/gnu/bin/i686-pc-linux-gnu-gcc -v output: Reading specs from /pkg/gnu/bin/../lib/gcc-lib/i686-pc-linux-gnu/3.2/specs
Configured with: ../../gcc-3.2/configure --prefix=/pkg/gnu/gcc/3.2
Thread model: posix
gcc version 3.2
2020-06-15T02:12:59.709Z [debug] [kit] Bad GCC binary /pkg/gnu/bin/gcc -v output: Reading specs from /pkg/gnu/bin/../lib/gcc-lib/i686-pc-linux-gnu/3.2/specs
Configured with: ../../gcc-3.2/configure --prefix=/pkg/gnu/gcc/3.2
Thread model: posix
gcc version 3.2

2020-06-15T02:12:59.710Z [debug] [kit] Found 0 kits in directory /pkg/gnu/bin
2020-06-15T02:12:59.710Z [debug] [kit] Detected GCC compiler: /usr/bin/X11/gcc-4.8
2020-06-15T02:12:59.710Z [debug] [kit] Detected GCC compiler: /usr/bin/X11/gcc
2020-06-15T02:12:59.711Z [debug] [kit] Found 2 kits in directory /usr/bin
2020-06-15T02:12:59.711Z [debug] [kit] Found 2 kits in directory /usr/bin/X11 **_--> after this message there is no further progress made and the snapshot below is shown_** 



**snapshot where the operation is stuck **
![image](https://user-images.githubusercontent.com/2882728/84612545-cc498b80-ae75-11ea-9e43-67368b0e39cb.png)

**Comment by sourabh2902** [comment]
Hi 
Is there any more information needed on this issue from my side ?

**Comment by andreeis** [comment]
Not for now. But we are curious to know whether CMake Tools 1.4.1 fixed the problem. Please confirm either way.

**Comment by sourabh2902** [comment]
HI @andreeis , 
Sure ! let me try the extension's latest version and see if the issue is still there. 

Thanks

**Comment by sourabh2902** [comment]
HI @andreeis 
I tried the latest version of the extension i.e. 1.4.1 
Please take a look at my steps and observations below: 

1. Opened VSCODE 
2. Uninstalled version 1.3.0 and installed version 1.4.1 ( I don't have any cmake-kits.json file either in .vscode/ or in /usr/local/share/CmakeTools/ directory.)
3. Opened the workspace on VSCODE. 
4. The extension kept on scanning for kits which never finished. --> just like the version 1.4.0
5. But there is an improvement this time, version 1.4.1 allows me to select a variant->configure-> build the code even when the kit scanning was going on. 
6. I waited for more than 2hours the kit scanning has not finished and looks to be showing the same symptoms like in the version 1.4.0

**Comment by andreeis** [comment]
Interesting. Unfortunately we still can't reproduce this and we didn't receive any other similar reports. I wonder what can be so different on your side. 

Are you unblocked now, though? Can you do all the normal development operations while ignoring the bottom right notification? Can the notification be closed?

Do you have another machine you can try this on?

Do you confirm that even after step 6, you still don't have any kits json file anywhere? After you confirm whether the kits file is there or not, can you run the usual command "Scan for kits" and let us know if that finishes and if that produces anything?

For the last resort, I can give you some instructions for how to clone, build and debug the extension on your side, just enough to give us a hint at least.

**Comment by sourabh2902** [comment]
Hi @andreeis 
Kindly take a look at my response below :

 _Are you unblocked now, though?_
Yes, I can select the variant, configure the project, build and debug the project using breakpoints while the silent scanning is going on in the background. Unlike version 1.4.0 the silent scanning step is not stopping me from carrying out my development operations anymore, so definitely there is some improvement. :) 

 _Can you do all the normal development operations while ignoring the bottom right notification? Can the notification be closed?_
Yes I am able to minimize this notification and able to continue

_Do you have another machine you can try this on?_
Actually all our codebase is IP sensitive, so I have tried the extension on remote Linux machines in IP protected area. I can not test this on my local machine because I don't have the code and necessary setup in local machine. 

_Do you confirm that even after step 6, you still don't have any kits json file anywhere?_ 
Yes I did the following : 
1, opened vscode, 
2. opened workspace
3. let the extension's silent scanning run for 12 hours 
4. Until now I dont see any kits file in any of these locations : 
./vscode
/usr/.config/Code/
/usr/.local/share/CmakeTools
/usr/.vscode/extensions 
Is there someplace that I should check for ? I also checked in these areas they have access permissions set to chmod - 770 

_After you confirm whether the kits file is there or not, can you run the usual command "Scan for kits" and let us know if that finishes and if that produces anything?_
I tried this also with exact same results as I mentioned above. The issue is tool starts scanning automatically as soon as I open the workspace, so I am not sure that when I invoke "Scan for kits", it actually does something or not. Please let me know when exactly should I invoke this step ?

**Comment by mateuszuram** [comment]
Hi All,

I'm seeing exactly the same problem as @sourabh2902 described with my setup.
Everything works fine with version 1.3.1, 1.4.0 is getting stuck on Scanning for kits, 1.4.1 and newer also getting stuck but I can run cmake configure, build and generally use a tool just all the time seeing windows in right bottom corner "Scanning for kits: ...".
In my case the log file, ~/.local/share/CMakeTools/log.txt, ends like this:
2020-11-13T18:55:35.463Z [debug] [driver] Setting new variant , Disable optimizations - include debug information.
2020-11-13T18:55:35.463Z [debug] [driver] Run _refreshExpansions
2020-11-13T18:55:35.463Z [debug] [driver] Run _refreshExpansions cb
2020-11-13T18:55:35.472Z [debug] [driver] Runnnig pre-configure checks and steps

Let me know if you would like me to run some tests or share some additional information.

OS: Linux
    Distributor ID: Ubuntu
    Description:    Ubuntu 16.04.6 LTS
    Release:        16.04
    Codename:       xenial
CMake Version: 3.16.2

**Comment by mateuszuram** [comment]
Hi,

I figured out my problem :).
In my company, CAD team is setting up wrapper scripts for CAD tools, the way they work is that they check environment variable for specific CAD tool version and execute it. If environment variable doesn't exists the script will block and wait for user to type the version he want's to run. One of those script was setup for xilinx edk power pc gcc compiler "Powerpc-eabi-gcc", when cmake extension executed this script it never returned and block entire scanning. After I removed those scripts from PATH everything works.
It might be good to add some timeout while executing compiler binary during scanning in the extension.

Regards,
Mat

**Comment by andreeis** [comment]
That explains it, thank you for the idea, it sounds good.

**Comment by sourabh2902** [comment]
HI @bobbrow  can you please tell me if this timeout feature has been added

**Comment by bobbrow** [comment]
It has not yet.

**Comment by sourabh2902** [comment]
@bobbrow  
Thanks for the response! 
Two years ago I reported this issue, at that time we couldn't get a resolution hence I have been using the older version 1.3.0 since then. A few weeks ago I tried updating to the latest version of vs code and Cmake tools and faced the issue. 
If you can share when this timeout feature can be added then it will be very helpful until then I will be forced to use older versions

**Comment by bobbrow** [comment]
I'm sorry that we haven't been able to do this yet.  We have 447 issues (as of this writing) and not as many developers working on this extension so it can be tough to get everything done that we would like.  This issue slipped under the radar just due to the volume of issues, but I put it in the list to consider for 1.11.  We also accept contributions from the community, so if you or anyone else out there would like to take a crack at it, we'd be happy to review your code and merge it in.

**Comment by sourabh2902** [comment]
Thanks @bobbrow !
I will wait for the next version release then. 

I noticed something: 
the $PATH in our environment is defined as : <some _path>/gnu64/gcc/7.2.0/bin 
Inside this path we have the following  : 

c++
gcc
gcc-ranlib  
gcov-tool                
x86_64-pc-linux-gnu-g++       
x86_64-pc-linux-gnu-gcc-ar     
 x86_64-pc-linux-gnu-gfortran
cpp  gcc-ar  gcov        
gfortran                
 x86_64-pc-linux-gnu-gcc       
 x86_64-pc-linux-gnu-gcc-nm
g++  
gcc-nm 
 gcov-dump   
x86_64-pc-linux-gnu-c++ 
 x86_64-pc-linux-gnu-gcc-7.2.0  
x86_64-pc-linux-gnu-gcc-ranlib


Out of this list we are interested in : 
g++ 
gcc 

I noticed that the latest tool version always goes into a infinite loop  for the following : 
x86_64-pc-linux-gnu-c++ 
 x86_64-pc-linux-gnu-gcc-7.2.0
x86_64-pc-linux-gnu-g++ 

Is there a way we can see that until this infinite loop is hit which kits were successfully scanned ? 
OR 
Is there a way I can mention a camke-kits.json and force the extension to use that even for first time starttup( not to allow the extension to scan for kits automatically for first time startup)   ?

**Comment by bobbrow** [comment]
Are there any circular symlinks there?  (can you do `ls -l` so we can see?)

**Comment by sourabh2902** [comment]
HI @bobbrow  
I checked the path actually there are no symlinks in the directory
Also, the same path is working for extension version 1.3.0 
In the latest version of the extension, The same path is getting used. 
In the latest version, I am able to do CMake configure and compile the code, just that scanning of kits never gets over

**Comment by snehith-shenoy** [comment]
Hi @bobbrow,

I can confirm the same issue from my side too. (CMake tools version 1.10.5)
I haven't tried the older version to get around this (unlike @sourabh2902 ) - I've been letting the extension do its futile scan in the background for now.

> Scanning for kits: Getting GCC version for /pkg/qct/software/gnu64/gcc/6.3.0/bin/x86_64-pc-linux-gnu-gcc

I have **/pkg/qct/software/gnu64/gcc/6.3.0/bin/** on PATH and that directory has the following files:
```
c++
cpp
g++
.............
gcc
gcc-ar
gcc-nm
gcc-ranlib
............
x86_64-pc-linux-gnu-c++
x86_64-pc-linux-gnu-g++
x86_64-pc-linux-gnu-gcc
x86_64-pc-linux-gnu-gcc-6.3.0
x86_64-pc-linux-gnu-gcc-ar
x86_64-pc-linux-gnu-gcc-nm
x86_64-pc-linux-gnu-gcc-ranlib
x86_64-pc-linux-gnu-gcj
x86_64-pc-linux-gnu-gfortran

```
There are no symlinks (all of the above are files). 
My suspicion is the tool looks for all files containing "gcc" although we're only interested in gcc & g++ in this case.

Running 
> /pkg/qct/software/gnu64/gcc/6.3.0/bin/x86_64-pc-linux-gnu-gcc -v

outputs:

> Using built-in specs.
> COLLECT_GCC=/pkg/qct/software/gnu64/gcc/6.3.0/bin/x86_64-pc-linux-gnu-gcc
> COLLECT_LTO_WRAPPER=/afs/localcell/cm/gv2.6/sysname/pkg.@sys/qct/software/gnu/gcc/6.3.0/bin/../libexec/gcc/x86_64-pc-linux-gnu/6.3.0/lto-wrapper
> Target: x86_64-pc-linux-gnu
> Configured with: ../configure --prefix=/prj/qct/sysadmin/sw_installs/suresh/gcc-6.3-install
> Thread model: posix
> gcc version 6.3.0 (GCC) 

Another suspicion is if something is going wrong with the version output parsing.

**Comment by bobbrow** [comment]
I put out a PR for this.  If it is due to one of the compilers never exiting (my best guess), it will print the command to the output window.  Timeout is 30 seconds.  I kept it high to avoid any possible false positives (e.g. slow VM).  If this goes in today, it will show up in tomorrow morning's pre-release.

**Comment by bobbrow** [comment]
Timeouts were added in 1.11.7 (pre-release).  If the timeout happens it should log the command line that hung.  `cmake.loggingLevel` needs to be "warning", "debug", or "info".  If you could try it out and let us know what you find, that would help us understand the root cause.

**Comment by sourabh2902** [comment]
Thanks @bobbrow  Kindly give me 1-2 days time I will use the latest versions and give you the feedback.

**Comment by snehith-shenoy** [comment]
Hey @bobbrow ,

I tried the latest pre-release version (1.11.7) with VS code:

> Version: 1.65.0
> Commit: b5205cc8eb4fbaa726835538cd82372cc0222d43
> Date: 2022-03-02T11:12:36.248Z
> Electron: 13.5.2
> Chromium: 91.0.4472.164
> Node.js: 14.16.0
> V8: 9.1.269.39-electron.0
> OS: Linux x64 4.12.14-122.113-default

I'm still seeing the same issue as before.
Here's the [CMakeTools.log](https://github.com/microsoft/vscode-cmake-tools/files/8450065/CMakeTools.log). (Note: I have masked the sensitive paths/ environment variable outputs;  let me know if you want any info)

**Comment by bobbrow** [comment]
Hi @snehith-shenoy,

It appears that this one never completes:
```
Testing GCC binary: /pkg/icetools/bin/Powerpc-eabi-gcc
Executing command: /pkg/icetools/bin/Powerpc-eabi-gcc -v
```
I'm not sure why the timeout didn't work.  What happens when you run that command in the terminal?  Where do you get that compiler from?

**Comment by bobbrow** [comment]
Sorry, my mistake.  I set the pull request to auto complete and assumed that the timeouts had been committed yesterday.  It seems some other change went in first and blocked my auto complete because it had to re-merge with main.  My change is going in today and will be available in Monday morning's pre-release.

**Comment by snehith-shenoy** [comment]
Hey @bobbrow,

> Sorry, my mistake. I set the pull request to auto complete and assumed that the timeouts had been committed yesterday. It seems some other change went in first and blocked my auto complete because it had to re-merge with main. My change is going in today and will be available in Monday morning's pre-release.

I tried out the latest pre-release and it looks like the timeout is **working**. Thanks for the fix


> Executing command:  /pkg/icetools/bin/Powerpc-eabi-gcc -v
> I'm not sure why the timeout didn't work. What happens when you run that command in the terminal? Where do you get that compiler from?

With latest version, the logs are something like this:

> [proc] Executing command: /pkg/icetools/bin/Powerpc-eabi-gcc -v
> [proc] Executing command: /pkg/gnu/bin/gcc -v
> [proc] Executing command: /pkg/gnu/bin/i686-pc-linux-gnu-gcc -v
> [proc] Executing command: /usr/bin/gcc -v
> [proc] Executing command: /usr/bin/gcc-4.8 -v
> [proc] Executing command: /usr/bin/X11/gcc -v
> [proc] Executing command: /usr/bin/X11/gcc-4.8 -v
> [proc] Executing command: /pkg/qct/software/gnu64/gcc/6.3.0/bin/x86_64-pc-linux-gnu-gcc -v
> [proc] **The command timed out: /pkg/icetools/bin/Powerpc-eabi-gcc -v**

And as far as the tool(Powerpc-eabi-gcc) is concerned, I'm pretty sure it isn't compiler as that command says -v isn't valid argument and displays some tool help when tried on command line manually.

This also adds to [my earlier guess ](https://github.com/microsoft/vscode-cmake-tools/issues/1289#issuecomment-1088411333) whether CMakeTools is just scanning for **all** binaries with "gcc" in their name(?); and if yes, is there any better way to pick binaries in the first place

**Comment by bobbrow** [comment]
> I tried out the latest pre-release and it looks like the timeout is **working**. Thanks for the fix

yay! 🥳 


> This also adds to [my earlier guess](https://github.com/microsoft/vscode-cmake-tools/issues/1289#issuecomment-1088411333) whether CMakeTools is just scanning for **all** binaries with "gcc" in their name(?); and if yes, is there any better way to pick binaries in the first place

Yes, that is one of the strategies we use to locate compiler candidates.  We only look for compilers that live in folders mentioned in the $PATH variable though.  We don't scan the whole file system.  We're open to other scanning ideas if you have any suggestions.
